synapse/lib/cell.py

```diff
@@ -2587,6 +2583,95 @@ async def _initCellBoot(self):
         async with s_telepath.loadTeleCell(self.dirn):
             await self._bootCellMirror(provconf)
 
+    @classmethod
+    async def _initBootRestore(cls, dirn):
+
+        env = 'SYN_RESTORE_HTTPS_URL'
+        rurl = os.getenv(env, None)
+
+        if rurl is None:
+            return
+
+        dirn = s_common.gendir(dirn)
+
+        # restore.done - Allow an existing URL to be left in the configuration
+        # for a service without issues.
+        doneiden = None
+
+        donepath = s_common.genpath(dirn, 'restore.done')
+        if os.path.isfile(donepath):
+            with s_common.genfile(donepath) as fd:
+                doneiden = fd.read().decode().strip()
+
+        rurliden = s_common.guid(rurl)
+
+        if doneiden == rurliden:
+            return
+
+        # If there is a mismatch vs restore.done, we hulk smash in the restored contents.
+        # There are no seatbelts here to check for existing files / etc.
+
+        rnfo = s_urlhelp.chopurl(rurl)
+
+        clean_url = s_urlhelp.sanitizeUrl(s_urlhelp.zipurl(rnfo))
+        logger.info(f'Restoring {cls.getCellType()} from {env}={clean_url}')
+
+        # Setup get args
+        rnfo_query = rnfo.pop('query', {})
+        # DISCUSS - SHOULD WE POP THIS OUT OF THE URL ENTIRELY IF PRESENT?
+        syn_ssl_verify = rnfo_query.get('syn_ssl_verify')
+        kwargs = {}
+        if str(syn_ssl_verify).lower() in ('0', 'false'):
+            kwargs['ssl'] = False
+
+        tmppath = s_common.gendir(dirn, 'tmp')
+        tarpath = s_common.genpath(tmppath, f'restore_{rurliden}.tgz')
+
+        try:
+
+            # DISCUSS
+            #
+            # THIS OPERATION CAN BE ALOT OF DEAD TIME FOR A LARGE BACKUP.
+            # WE COULD print EVERY 1% DOWNLOADED PER CHUNK IF WE HAVE
+            # A content-length RESPONSE HEADER?
+            with s_common.genfile(tarpath) as fd:
+                async with aiohttp.client.ClientSession() as sess:
+                    async with sess.get(rurl, **kwargs) as resp:
+                        resp.raise_for_status()
+                        csize = s_const.megabyte * 8  # arbitrary magic number
+                        async for chunk in resp.content.iter_chunked(csize):
+                            fd.write(chunk)
+
+            logger.info(f'Extracting {tarpath} to {dirn}')
+
+            with tarfile.open(tarpath) as tgz:
+                for memb in tgz.getmembers():
+                    if memb.name.find('/') == -1:
```

Comment:
this check and the conversion below feel a bit fragile. Are we just trying to confirm a leading `/` and strip it?

Reply:
This is the same convention we use when we pull from backups during provisioning.

Reply:
Looking at adding a manifest at least to give some metadata about the backup inside of it; #2867 is open against this PR.